File: go/aws/transport_sqs_test.go
package aws

import (
	"context"
	"encoding/json"
	"errors"
	"testing"

	"github.com/aws/aws-lambda-go/events"

	"github.com/sns45/anyhook/go/core"
)

func makeSqsEvent(t *testing.T, records []struct {
	messageID string
	m         core.Message
}) events.SQSEvent {
	t.Helper()
	out := events.SQSEvent{Records: make([]events.SQSMessage, len(records))}
	for i, r := range records {
		body, err := json.Marshal(r.m)
		if err != nil {
			t.Fatal(err)
		}
		out.Records[i] = events.SQSMessage{MessageId: r.messageID, Body: string(body)}
	}
	return out
}

func TestSqsTransport_Send(t *testing.T) {
	ddb := newFakeDynamo()
	sqsClient := newFakeSQS()
	env := testEnv(ddb, sqsClient)
	transport := NewSqsTransport(env)

	m := makeTestMessage(nil)
	if err := transport.Send(context.Background(), m); err != nil {
		t.Fatal(err)
	}

	calls := sqsClient.calls()
	if len(calls) != 1 {
		t.Fatalf("SendMessage calls = %d, want 1", len(calls))
	}
	if *calls[0].QueueUrl != env.QueueURL {
		t.Errorf("QueueUrl = %q, want %q", *calls[0].QueueUrl, env.QueueURL)
	}
	var sent core.Message
	if err := json.Unmarshal([]byte(*calls[0].MessageBody), &sent); err != nil {
		t.Fatal(err)
	}
	if sent.MessageID != m.MessageID {
		t.Errorf("sent.MessageID = %q, want %q", sent.MessageID, m.MessageID)
	}
}

func TestSqsTransport_SendReusesClientAcrossCalls(t *testing.T) {
	ddb := newFakeDynamo()
	sqsClient := newFakeSQS()
	transport := NewSqsTransport(testEnv(ddb, sqsClient))

	if err := transport.Send(context.Background(), makeTestMessage(nil)); err != nil {
		t.Fatal(err)
	}
	if err := transport.Send(context.Background(), makeTestMessage(nil)); err != nil {
		t.Fatal(err)
	}
	if len(sqsClient.calls()) != 2 {
		t.Fatalf("SendMessage calls = %d, want 2", len(sqsClient.calls()))
	}
}

func TestSqsTransport_DispatchRecordsMapsEachRecord(t *testing.T) {
	transport := NewSqsTransport(testEnv(newFakeDynamo(), newFakeSQS()))
	var seen []string
	if err := transport.Subscribe(context.Background(), func(_ context.Context, m core.Message) error {
		seen = append(seen, m.MessageID)
		return nil
	}); err != nil {
		t.Fatal(err)
	}

	m1 := makeTestMessage(func(m *core.Message) { m.MessageID = "good-1" })
	m2 := makeTestMessage(func(m *core.Message) { m.MessageID = "good-2" })
	event := makeSqsEvent(t, []struct {
		messageID string
		m         core.Message
	}{
		{"r1", m1}, {"r2", m2},
	})

	res, err := transport.DispatchRecords(context.Background(), event, nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(seen) != 2 || seen[0] != "good-1" || seen[1] != "good-2" {
		t.Fatalf("seen = %v, want [good-1 good-2]", seen)
	}
	if len(res.BatchItemFailures) != 0 {
		t.Errorf("BatchItemFailures = %v, want empty", res.BatchItemFailures)
	}
}

func TestSqsTransport_DispatchRecordsPartialBatchFailure(t *testing.T) {
	transport := NewSqsTransport(testEnv(newFakeDynamo(), newFakeSQS()))
	var seen []string
	if err := transport.Subscribe(context.Background(), func(_ context.Context, m core.Message) error {
		seen = append(seen, m.MessageID)
		if m.MessageID == "bad" {
			return errors.New("delivery blew up")
		}
		return nil
	}); err != nil {
		t.Fatal(err)
	}

	event := makeSqsEvent(t, []struct {
		messageID string
		m         core.Message
	}{
		{"r1", makeTestMessage(func(m *core.Message) { m.MessageID = "good" })},
		{"r2", makeTestMessage(func(m *core.Message) { m.MessageID = "bad" })},
		{"r3", makeTestMessage(func(m *core.Message) { m.MessageID = "good-2" })},
	})

	res, err := transport.DispatchRecords(context.Background(), event, nil)
	if err != nil {
		t.Fatal(err)
	}
	// Every record was attempted; the batch did not short-circuit.
	if len(seen) != 3 || seen[0] != "good" || seen[1] != "bad" || seen[2] != "good-2" {
		t.Fatalf("seen = %v, want [good bad good-2]", seen)
	}
	if len(res.BatchItemFailures) != 1 || res.BatchItemFailures[0].ItemIdentifier != "r2" {
		t.Fatalf("BatchItemFailures = %v, want [{ItemIdentifier:r2}]", res.BatchItemFailures)
	}
}

func TestSqsTransport_DispatchRecordsExplicitHandlerOverridesSubscribed(t *testing.T) {
	transport := NewSqsTransport(testEnv(newFakeDynamo(), newFakeSQS()))
	var subscribed, explicit []string
	if err := transport.Subscribe(context.Background(), func(_ context.Context, m core.Message) error {
		subscribed = append(subscribed, m.MessageID)
		return nil
	}); err != nil {
		t.Fatal(err)
	}

	event := makeSqsEvent(t, []struct {
		messageID string
		m         core.Message
	}{
		{"r1", makeTestMessage(func(m *core.Message) { m.MessageID = "explicit" })},
	})
	res, err := transport.DispatchRecords(context.Background(), event, func(_ context.Context, m core.Message) error {
		explicit = append(explicit, m.MessageID)
		return nil
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(subscribed) != 0 {
		t.Errorf("subscribed handler was invoked = %v, want it bypassed by the explicit handler", subscribed)
	}
	if len(explicit) != 1 || explicit[0] != "explicit" {
		t.Errorf("explicit = %v, want [explicit]", explicit)
	}
	if len(res.BatchItemFailures) != 0 {
		t.Errorf("BatchItemFailures = %v, want empty", res.BatchItemFailures)
	}
}

func TestSqsTransport_DispatchRecordsBeforeSubscribeErrors(t *testing.T) {
	transport := NewSqsTransport(testEnv(newFakeDynamo(), newFakeSQS()))
	_, err := transport.DispatchRecords(context.Background(), events.SQSEvent{}, nil)
	if err == nil {
		t.Fatal("expected an error when dispatching before Subscribe with no explicit handler")
	}
}

File: packages/aws/test/transport-sqs.test.ts
/** @fileoverview SqsTransport: send() composes the @anyq/sqs producer; dispatchRecords maps an SQS batch to per-record deliveries with partial-batch-failure reporting. @module @anyhook/aws (test) */
import { describe, test, expect, beforeAll, afterAll } from 'bun:test';
import { createHash } from 'node:crypto';
import type { SQSEvent } from 'aws-lambda';
import type { Message } from '@anyhook/core';
import { SqsTransport } from '../src/transport-sqs.js';

// NOTE on why this file does NOT use aws-sdk-client-mock for `send()`: `@anyq/sqs` ships a fully
// bundled `dist/index.js` (its own build inlines `@aws-sdk/client-sqs` rather than keeping it
// external), so the `SQSClient` class it constructs internally is a DIFFERENT class object than the
// one this test could import from `@aws-sdk/client-sqs` — `aws-sdk-client-mock` patches a specific
// class's prototype, so it cannot see calls made through that bundled copy. `state-dynamo.test.ts`
// and `scheduler.test.ts` use `aws-sdk-client-mock` freely because `state-dynamo.ts`/`scheduler.ts`
// construct the AWS SDK clients THIS package imports directly (an injectable `Env`). Here we instead
// run a real local HTTP server standing in for the SQS JSON-protocol endpoint (same technique
// `../../cloudflare/test/http-fetch.test.ts` uses for `fetch`) and assert on the real request it received.

let sqsServer: ReturnType<typeof Bun.serve>;
let received: { body: string }[] = [];

beforeAll(() => {
  process.env.AWS_ACCESS_KEY_ID ??= 'test';
  process.env.AWS_SECRET_ACCESS_KEY ??= 'test';
  sqsServer = Bun.serve({
    port: 0,
    async fetch(req) {
      const body = await req.text();
      received.push({ body });
      const md5 = createHash('md5').update((JSON.parse(body) as { MessageBody: string }).MessageBody).digest('hex');
      return new Response(JSON.stringify({ MD5OfMessageBody: md5, MessageId: `sqs-${received.length}` }), {
        status: 200,
        headers: { 'content-type': 'application/x-amz-json-1.0' },
      });
    },
  });
});

afterAll(() => sqsServer.stop(true));

function endpoint(): string {
  return `http://127.0.0.1:${sqsServer.port}`;
}

function makeMessage(overrides: Partial<Message> = {}): Message {
  return {
    messageId: `msg_${crypto.randomUUID()}`,
    tenant: 'tenant-a',
    endpointId: 'ep_1',
    eventId: `evt_${crypto.randomUUID()}`,
    eventType: 'payment.succeeded',
    payload: { hello: 'world' },
    attemptNo: 0,
    status: 'pending',
    createdAt: Date.now(),
    ...overrides,
  };
}

function makeSqsEvent(records: { messageId: string; body: string }[]): SQSEvent {
  return {
    Records: records.map((r) => ({
      messageId: r.messageId,
      receiptHandle: `rh-${r.messageId}`,
      body: r.body,
      attributes: {
        ApproximateReceiveCount: '1',
        SentTimestamp: '0',
        SenderId: 'test',
        ApproximateFirstReceiveTimestamp: '0',
      },
      messageAttributes: {},
      md5OfBody: 'ignored',
      eventSource: 'aws:sqs',
      eventSourceARN: 'arn:aws:sqs:us-east-1:123456789:anyhook',
      awsRegion: 'us-east-1',
    })),
  } as unknown as SQSEvent;
}

describe('SqsTransport.send: composes the @anyq/sqs producer', () => {
  test('publishes the message onto the configured queue', async () => {
    received = [];
    const queueUrl = `${endpoint()}/123456789/anyhook`;
    const transport = new SqsTransport({ queueUrl, region: 'us-east-1', endpoint: endpoint() });
    const m = makeMessage();
    await transport.send(m);

    expect(received.length).toBe(1);
    const sent = JSON.parse(received[0]!.body) as { QueueUrl: string; MessageBody: string };
    expect(sent.QueueUrl).toBe(queueUrl);
    expect(JSON.parse(sent.MessageBody)).toEqual(m);
  });

  test('reuses the connected producer across multiple sends (connects once)', async () => {
    received = [];
    const transport = new SqsTransport({ queueUrl: `${endpoint()}/123456789/anyhook`, endpoint: endpoint() });
    await transport.send(makeMessage());
    await transport.send(makeMessage());
    expect(received.length).toBe(2);
  });
});

describe('SqsTransport.dispatchRecords: per-record delivery + partial-batch-failure reporting', () => {
  test('maps each record through the subscribed handler', async () => {
    const transport = new SqsTransport({ queueUrl: 'https://q' });
    const seen: string[] = [];
    await transport.subscribe(async (m) => {
      seen.push(m.messageId);
    });

    const m1 = makeMessage({ messageId: 'good-1' });
    const m2 = makeMessage({ messageId: 'good-2' });
    const event = makeSqsEvent([
      { messageId: 'r1', body: JSON.stringify(m1) },
      { messageId: 'r2', body: JSON.stringify(m2) },
    ]);

    const res = await transport.dispatchRecords(event);
    expect(seen).toEqual(['good-1', 'good-2']);
    expect(res.batchItemFailures).toEqual([]);
  });

  test('a throwing record is reported in batchItemFailures; the rest of the batch is unaffected', async () => {
    const transport = new SqsTransport({ queueUrl: 'https://q' });
    const seen: string[] = [];
    await transport.subscribe(async (m) => {
      seen.push(m.messageId);
      if (m.messageId === 'bad') throw new Error('delivery blew up');
    });

    const event = makeSqsEvent([
      { messageId: 'r1', body: JSON.stringify(makeMessage({ messageId: 'good' })) },
      { messageId: 'r2', body: JSON.stringify(makeMessage({ messageId: 'bad' })) },
      { messageId: 'r3', body: JSON.stringify(makeMessage({ messageId: 'good-2' })) },
    ]);

    const res = await transport.dispatchRecords(event);
    expect(seen).toEqual(['good', 'bad', 'good-2']); // every record was attempted; the batch did not short-circuit
    expect(res.batchItemFailures).toEqual([{ itemIdentifier: 'r2' }]);
  });

  test('an explicit handler argument overrides the subscribed one', async () => {
    const transport = new SqsTransport({ queueUrl: 'https://q' });
    const seen: string[] = [];
    const event = makeSqsEvent([{ messageId: 'r1', body: JSON.stringify(makeMessage({ messageId: 'explicit' })) }]);
    const res = await transport.dispatchRecords(event, async (m) => {
      seen.push(m.messageId);
    });
    expect(seen).toEqual(['explicit']);
    expect(res.batchItemFailures).toEqual([]);
  });

  test('throws if dispatched before subscribe() and no handler is passed', async () => {
    const transport = new SqsTransport({ queueUrl: 'https://q' });
    await expect(transport.dispatchRecords(makeSqsEvent([]))).rejects.toThrow(/subscribe/);
  });
});
